app/lib/shared/utils.dart

```diff
@@ -361,3 +362,39 @@ bool fixedTimeEquals(String a, String b) {
   }
   return result == 0;
 }
+
+extension StreamBoundedForEach<T> on Stream<T> {
+  /// Executes items using [eachFn] with maximum [concurrency].
+  Future<void> boundedForEach(
+    int concurrency,
+    FutureOr<void> Function(T) eachFn,
+  ) async {
+    final pool = Pool(concurrency);
+    final futures = <Future>[];
+    try {
+      await for (final item in this) {
+        final f = pool.withResource(() => eachFn(item));
+        futures.add(f);
+        if (futures.length >= concurrency) {
+          await Future.wait(futures);
+          futures.clear();
+        }
+      }
```

Comment:
We could also use `Future.any()` and then find a way to remove the completed future from the list of `futures`.

But this is much harder and more expensive, there is really no reason to keep a list of all `futures`, that's just a lot of unnecessary allocations :D

Reply:
Given the alternative, I find the `Future.any` + removing the completed Future from the list more appealing. I think the overhead of an update in a small list is probably very little compared to the async processing we are doing, I'd rather keep the for-loop simple. Let's explore that first before going in for the semaphore version.

Reply:
We could make this nicer using a `Condition` like:
```dart
/// A [Condition] allows micro-tasks to [wait] for other micro-tasks to
/// [notify].
///
/// [Condition] is a concurrency primitive that allows one micro-task to
/// wait for notification from another micro-task. The [Future] return from
/// [wait] will be completed the next time [notify] is called.
final class Condition {
  var _completer = Completer<void>();

  /// Complete all futures previously returned by [wait].
  ///
  /// Calls to [wait] after this call, will not be resolved, until the next time
  /// [notify] is called.
  void notify() {
    if (!_completer.isCompleted) {
      _completer.complete();
    }
  }

  /// Returns a [Future] that will complete the next time [notify] is called.
  ///
  /// This will always return an unresolved [Future]. Once [notify] is called
  /// the future will be completed, and any new calls to [wait] will return a
  /// new future. This future will also be unresolved, until [notify] is called.
  ///
  /// The [Future] return from this condition will never throw.
  Future<void> get wait {
    if (_completer.isCompleted) {
      _completer = Completer();
    }
    return _completer.future;
  }
}

```